### ISSUE TITLE
Slightly different variations of clips after hitting "Replay" button

### ISSUE THREAD

**Issue by amy-wong-1** [charconst/homebank_flurry_frontend#10]
Hi everyone,
I've noticed that when I hit "replay" on some of the clips, I sometimes hear a slightly different variation of the clip. It seems to  cut off a bit of the audio at the end and/or add a little bit of audio at the end. I've attached a video that shows this happening. 

https://user-images.githubusercontent.com/88563470/128578640-db74a044-86e9-4d5b-a7ae-c61d93a1ed29.mp4

**Comment by charconst** [comment]
@amy-wong-1 Hi Amy, thanks for reporting this. Few questions to help troubleshoot what's happening:

(1) What web browser are you using?
(2) What clip # (s) did you notice this issue happening?
(3) When did you first notice this issue happening?
(4) Is the issue happening every time you click replay or only on certain clips?

**Comment by amy-wong-1** [comment]
Hi Charlie,

To answer your questions, I am using Safari to access HomeBank Flurry. I first noticed this happening today. Unfortunately, I do not specifically remember which clips I saw this happening on, but it has definitely been on several clips (I am currently on clip 800/2540, and I feel that it has happened many times so far). It appears to only be happening on certain clips from what I can tell.

**Comment by amy-wong-1** [comment]
Hi Charlie,

As I'm actively looking out for the issue, I've noticed it happening more often. I've noticed it on clips #800, 806, 813, 816, and 820, just to name a few.

**Comment by charconst** [comment]
I've tried to reproduce it on my end on three different browsers (including Safari) and I'm not having the issue. Can you try to log in with Chrome and see if any of those clips are giving you issues?

**Comment by charconst** [comment]
When I go to clip #800 in Chrome this is what I see/hear:
https://user-images.githubusercontent.com/44410850/128583575-b36cb5ac-ae07-4c4e-9f0f-423080e1fd71.mov

**Comment by amy-wong-1** [comment]
I'm on Chrome, and the issue does not seem to be happening as frequently. I don't believe there's an option for me to go back to those clips and check. I'm currently on #841 on Chrome, and I believe I hear a slight discrepancy. 

https://user-images.githubusercontent.com/88563470/128583854-e8d69ef9-8f97-4839-8760-db3c73a7bfec.mov

**Comment by charconst** [comment]
Yeah, I feel like it's related to the browser. I can reset you back to #800 for testing purposes. Would you like me to do that?

**Comment by AnneSWarlaumont** [comment]
Thank you both for your work on this!!

I can hear the discrepancies in both of Amy's videos. To me it sounds like the differences are toward the end of the clip. Can't tell if that is due to sound being cut off or being added—we could look into that locally to find out which though.

I am not sure if resetting back to #800 is worth it because it seems like this has been happening throughout and the inputs Amy has been giving are generally still good data (probably the errors introduced are not worse than the ones that happen sometimes just from mishearing, or making an occasional mistake). It could be more useful for Amy to continue sharing notes on which clips it is and isn't happening on and what the nature of the discrepancies are or if for any of the clips the discrepancy would effect the decision (1–5 judgment) for the clip. And perhaps next time it happens, Amy can pause on that clip and let us know that this would be a good opportunity to try out some different things?

**Comment by amy-wong-1** [comment]
That works for me. Going forward, I will continue to use Chrome to access HomeBank Flurry and will make note of any discrepancies that I come across. Thank you all for your help!

**Comment by amy-wong-1** [comment]
Just wanted to share this video of the issue that I did not get a chance to upload (this was taken when I was still using Safari to access HomeBank Flurry). For this clip, I was having trouble making a judgment due to the discrepancy. 

https://user-images.githubusercontent.com/88563470/128805336-c2d57e76-e34d-4757-adfd-1192ae327a41.mp4

**Comment by amy-wong-1** [comment]
I'm currently on clip 1480 on Chrome, and I believe I hear a discrepancy that affects my judgment for labeling the clip. I will pause on this clip for now.. 
https://user-images.githubusercontent.com/88563470/128809992-c77e4127-c870-4c02-ab3d-abb219a3dbf7.mp4

**Comment by amy-wong-1** [comment]
Also, I just want to add that fortunately, I've seen significantly less instances of discrepancies while using Chrome in comparison to Safari. The one above is the only instance I've seen since the first instance.

**Comment by amy-wong-1** [comment]
I'm currently on clip #2010 on Chrome, and I'm hearing a discrepancy that affects my judgment. I also heard the discrepancy with clip #1574, but that one did not make a difference in my judgment. I've attached a video of clip #2010. 

https://user-images.githubusercontent.com/88563470/129377550-ee3fe3b2-0469-475f-88db-cd072abb5e9c.mp4

**Comment by amy-wong-1** [comment]
Here is another instance of a discrepancy in clip #2097.

https://user-images.githubusercontent.com/88563470/129381534-0e5f66ee-4949-4bd0-a869-4d4a0fe9748b.mp4

**Comment by charconst** [comment]
I've pushed a fix live for this. It uses a higher resolution timer to determine when to stop the audio clip. Let me know if that helps.

**Comment by amy-wong-1** [comment]
I'm currently working on file audio/0530_000304_scrubbed.wav and am on clip 110, and I encountered this discrepancy. Besides this one, the button seems to be working better.

https://user-images.githubusercontent.com/88563470/129997124-66e53508-ea2e-47ff-83aa-c5858f8d9893.mov

**Comment by amy-wong-1** [comment]
I also encountered this slight discrepancy on clip 130. 
https://user-images.githubusercontent.com/88563470/129997925-3fcda197-620e-4c5f-9409-ffa9f6cb5639.mp4

**Comment by amy-wong-1** [comment]
I'm not sure if this is a separate issue or if it is related to the updated "replay" button, but I'm noticing that now when I click "next clip," there often seems to be a shortened version of the clip that is automatically played. I have found that I need to click the "replay" button to ensure that I hear the full clip. Here is an example of this happening on clip #985.

https://user-images.githubusercontent.com/88563470/130152526-834ac439-9154-43b6-87b1-d982a3db3404.mov

**Comment by charconst** [comment]
I'm almost wondering if the HTML audio player API is introducing some latency. For example, if calls to .Pause() don't happen immediately on that frame (or deterministically) then it could introduce variable lag.

**Comment by AnneSWarlaumont** [comment]
Thanks so much, Amy, for noting these patterns/issues—It definitely is problematic, both inconvenient/confusing for the listener but also for the label quality we'll get. And thank you, Charlie, for those ideas.

Looking at AudioClipContainer.tsx I see what you're saying where playAudio should be doing the same thing for Next Clip vs. Replay Clip, but I don't understand the code enough at least at this point to understand what could be going on or what you're suggesting might be the problem. I do wonder if having Amy share console logs could help with debugging though.

An alternative approach could still be to not be loading up the whole audio file and playing from there. Instead we could work from smaller wav files, each representing the exact clip to be played. That might work better with the html audio player, but obviously would be a major restructuring and maybe not feasible given our current personnel resources. Also, in other projects I sometimes clip that smaller wav out from the larger file on the fly, saving it as a temp file, playing the temp file, then deleting/overwriting the temp file. Not sure if that is possible within this app though.
